lambda/transl_array_comprehension.ml

```diff
@@ -0,0 +1,932 @@
+           raise an error here instead *)
+
+  (** Compute the size of a single nonempty array iterator in the fixed-size
+      array case.  This is either the size of a range, which itself is either
+      [stop - start + 1] or [start - stop + 1] depending on if the array is
+      counting up ([to]) or down ([downto]), clamped to being nonnegative; or it
+      is the length of the array being iterated over.  In the range case, we
+      also have to check for overflow.  We require that the iterators be
+      nonempty, although this is only important for the range case; generate
+      Lambda code that checks the result of [are_any_empty] before entering
+      [size_nonempty] to ensure this. *)
+  let size_nonempty ~loc : Usage.many t -> lambda = function
+    | Range { start     = start
+            ; stop      = stop
+            ; direction = Reusable direction }
+      ->
+        let open (val Lambda_utils.int_ops ~loc) in
+        let start = Lvar start.id in
+        let stop  = Lvar stop.id in
+        let low, high = match direction with
+          | Upto   -> start, stop
+          | Downto -> stop,  start
+        in
+        (* We can assume that the range is nonempty, but computing its size
+           still might overflow *)
+        let range_size = Ident.create_local "range_size" in
+        Llet(Alias, Pintval, range_size, (high - low) + l1,
+          (* If the computed size of the range is positive, there was no
+             overflow; if it was zero or negative, then there was overflow *)
+          Lifthenelse(Lvar range_size > l0,
+            Lvar range_size,
+            Precompute_array_size.raise_overflow_exn ~loc,
+            Pintval))
+    | Array { iter_arr = _; iter_len = Reusable iter_len } ->
+        Lvar iter_len.id
+
+  (** Compute the total size of an array built out of a list of translated
+      iterators in the fixed-size array case, as long as all the iterators are
+      nonempty; since this forms a cartesian product, we take the product of the
+      sizes (see [size_nonempty]).  This can overflow, in which case we will
+      raise an exception.  This is the operation needed to precompute the fixed
+      size of a nonempty fixed-size array; check against [are_any_empty] first
+      to address the case of fixedly-empty array. *)
+  let total_size_nonempty ~loc (iterators : Usage.many t list) =
+    Precompute_array_size.safe_product_pos
+      ~variable_name:"iterator_size"
+      ~loc
+      (List.map (size_nonempty ~loc) iterators)
+end
+
+(** Machinery for working with resizable arrays for the results of an array
+    comprehension: they are created at a fixed, known, small size, and are
+    doubled in size when necessary.  These are the arrays that back array
+    comprehensions by default, but not in the fixed-size case; in that case, we
+    simply construct an array of the appropriate size directly. *)
+module Resizable_array = struct
```

Comment:
(Copied from the longer block, but I think we can discuss here.)  As we discussed, I think this isn't worth pulling out here, especially since we take some advantage of the fact that we're just monkeying with two different kinds of arrays.  If we find ourselves needing this elsewhere, then I'd revisit it.

Reply:
I don't really feel like it's worth just creating a new internal module for this, especially when we want to e.g. sometimes just run through an array; if we were trying to actually add one to the standard library, I'd be more open to considering it.  But feel free to try to convince me!

Reply:
OK. Maybe just leave a comment in the file stating that an alternative implementation would be to have the internal module, but avoiding it seemed simpler. Then a reader knows that there's nothing wrong with the alternative.